### DISCUSSION TITLE
Regarding dates ...

### DISCUSSION THREAD

**Discussion by MortenDHansen** [owen-it/laravel-auditing#799] [Ideas]
Hi all! (@erikn69 , @anteriovieira @parallels999 etc) 

I am preparing a v14 branch, and in that process i am considering different scenarios that has been handled "here and there". One of those are how we handle datetime. 

I am torn between letting the package itself lean on the principle of "simple storage of whatever is currently written", or "recording what is in the table". 

The difference i subtle, but with regards to dates, the timezone on save is a piece of information we are ignoring when we save the datetime string as we are used to seeing it in MySQL. 

Most database engines will keep a datetime column in some sort of format that you unpack on retrieval. As PHP and laravel developers we may be used to seeing the "Y-m-d H:i:s" format, but if we lean on this in the package, we may be limiting a bit too much, as this is not literally the information we have.

Laravels own json serializer formats according to the ISO-8601 standard (as proposed here: https://github.com/laravel/ideas/issues/1940) - and thus basically doing what carbon does

My suggestion would be to change behaviour to lean on the ISO-8601 approach when serializing dates to json. 

Pros or cons?

**Comment by MortenDHansen** [comment]
Basically, we currently takes "whatever eloquent casts to us". Im saying we should consider taking what is in the table instead.

**Reply by anteriovieira** [reply]
Hi @MortenDHansen , I agree with you.

**Comment by anteriovieira** [comment]
Hi, @erikn69 @parallels999 any thoughts?

And can you send me your [discord](https://discord.gg/nJcABK8H) username? Then I can invite you to private channels.

**Comment by parallels999** [comment]
> Pros or cons?

What about backwards compatibility? 
I mean with those who have millions of audits with "whatever eloquent casts to us"
Is there any way to upgrade those json?

**Reply by MortenDHansen** [reply]
I _believe_ that since the change will only have effect going forward, the existing records would be handled as they are now, but going forward they are saved different.

Meaning, if we saved "2012-05-05 10:15:22" that value will be unchanged, but going forward we would save "2012-05-05T10:15:22Z+0000", which in the end SHOULD amount to the same output given the application handles dates the same. 

Looking at this recent pull request, we are mitigating a shortcoming, and potentially harming old records: https://github.com/owen-it/laravel-auditing/pull/775/files#diff-d85e57ca553073b6b08abdded140ee91754ba3503bf62660ff2a44685a7156cbR153 

We look into the date casts, then we assume we have a date which we exposes as a date, but we have no timezone info to do it from :( - if we change the saved the timestamp, the application would work the same - convert to datetime, except now timestamp is read including the original timezone (which may not even be different). Cast is the same on output.

**Reply by parallels999** [reply]
in that case it looks good to me, 
it is better to simulate the operation of laravel as much as possible

**Reply by MortenDHansen** [reply]
Well, no guarantee i'm right, but the above pull request appears to be handling a case, that could have been handled correctly. I fear that its harder to maintain and reason fixes, than it is to solve the issue. I am a little uncertain about the effects of the change though.

**Reply by parallels999** [reply]
That's also a problem, it takes time to fix unexpected issues, possible incompatibilities with 3rd party packages using this

PHP7 support should be removed and PHP8 syntax adopted


Is https://github.com/owen-it/laravel-auditing/issues/635#issuecomment-1465152078, https://github.com/owen-it/laravel-auditing/issues/588 related with this?

**Reply by MortenDHansen** [reply]
I'll push the current state of my proposed v14, but yes. I aim to make the behavior consistent so that these issues would not occur. ... And drop php 7.x support following laravels minimum 8.0

**Reply by MortenDHansen** [reply]
Here is the v14 dev base branch. https://github.com/owen-it/laravel-auditing/tree/v14-dev

Note it's not a v14 branch (yet)!

**Comment by MortenDHansen** [comment]
Alright. I stabbed at dates, casts and serialization, but i'm having a hard time finding and proving actual issues. It's more or less "just output logic" 🤷 

Im dropping the serialization refactor all together 😄
